src/node/routes/login.ts

```diff
@@ -62,24 +61,31 @@ router.get("/", async (req, res) => {
 })
 
 router.post("/", async (req, res) => {
+  const password = sanitizeString(req.body.password)
+  const hashedPasswordFromArgs = req.args["hashed-password"]
+
   try {
     // Check to see if they exceeded their login attempts
     if (!limiter.canTry()) {
       throw new Error("Login rate limited!")
     }
 
-    if (!req.body.password) {
+    if (!password) {
       throw new Error("Missing password")
     }
 
-    if (
-      req.args["hashed-password"]
-        ? safeCompare(hash(req.body.password), req.args["hashed-password"])
-        : req.args.password && safeCompare(req.body.password, req.args.password)
-    ) {
+    const passwordMethod = getPasswordMethod(hashedPasswordFromArgs)
+    const { isPasswordValid, hashedPassword } = await handlePasswordValidation({
+      passwordMethod,
+      hashedPasswordFromArgs,
+      passwordFromRequestBody: password,
+      passwordFromArgs: req.args.password,
+    })
+
+    if (isPasswordValid) {
       // The hash does not add any actual security but we do it for
       // obfuscation purposes (and as a side effect it handles escaping).
-      res.cookie(Cookie.Key, hash(req.body.password), {
+      res.cookie(Cookie.Key, hashedPassword, {
```

Comment:
While hashing is a major step forward, the problem with this approach is that it still allows attackers who have access to the hash to just submit it as-is and gain access to code-server - effectively not very different from storing it in plaintext.

We should definitely look into replacing this with something more robust so that password hashing isn't just a placebo - now or in a future version is up to you, but IMO we should tackle this before asking users to upgrade to Argon2.

Reply:
> effectively not very different from storing it in plaintext.

I'm really glad you mention that and agree.

> We should definitely look into replacing this with something more robust so that password hashing isn't just a placebo

Also agree! 

> now or in a future version is up to you, but IMO we should tackle this before asking users to upgrade to Argon2

I would argue that it falls out of the scope of this PR. If we close this PR without merging, we're in the same situation we're discussing right now (see [code](https://github.com/cdr/code-server/blob/main/src/node/routes/login.ts#L82-L86)) because we're doing the same thing - hashing the password and storing as a cookie. 

The scope of this PR was to replace the hashing algorithm (not to fix the problem with attackers who have the hash).

Therefore, I think it makes sense to separate the two concerns and tackle them in separate PRs. I will create a new issue to track that though!

Reply:
https://github.com/cdr/code-server/issues/3576

Reply:
Totally get it. Sometimes PRs bring in a lot of heat just because they touch an area that would otherwise have a lot of attention drawn towards. This one specifically has allowed us to become aware of other things in the codebase we want to change.

:pray: